File: js/test/kms/historyApi.test.ts
import process from 'process';

import { MapleStoryApiError, MapleStoryApiErrorCode } from '../../src';
import { MapleStoryApi } from '../../src/maplestory/api/kms';
import { toString } from '../utils';

const apiKey = process.env.API_KEY_KMS!; // Your API Key
const api = new MapleStoryApi(apiKey);

describe('History Information Retrieval', () => {
  describe('getStarforceHistory', () => {
    test('success: getStarforceHistory', async () => {
      const count = 1000;
      const response = await api.getStarforceHistory(count);
      console.log(toString(response));
    });

    test('success: getStarforceHistory with date', async () => {
      const count = 1000;
      const response = await api.getStarforceHistory(count, {
        year: 2025,
        month: 9,
        day: 21,
      });
      console.log(toString(response));
    });

    test('success: getStarforceHistory with cursor', async () => {
      const count = 10;
      const response = await api.getStarforceHistory(count, {
        year: 2025,
        month: 9,
        day: 21,
      });
      const { nextCursor } = response; // 결과가 count 보다 많은 경우에만 nextCursor가 반환 됩니다.
      expect(nextCursor).toBeDefined();
      const nextResponse = await api.getStarforceHistory(count, nextCursor!);
      console.log(toString(nextResponse));
    });

    test('fail: getStarforceHistory with invalid date', async () => {
      try {
        await api.getStarforceHistory(1000, {
          year: 2023,
          month: 12,
          day: 26,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2023-12-27.');
        console.log(error.message);
      }
    });

    test('fail: getStarforceHistory with invalid cursor throw OPENAPI00003', async () => {
      const invalidCursor = 'invalid_cursor_123';
      try {
        await api.getStarforceHistory(1000, invalidCursor);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });

    test('fail: getStarforceHistory with excessive count throw OPENAPI00004', async () => {
      try {
        const count = 9999;
        await api.getStarforceHistory(count);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00004);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getPotentialHistory', () => {
    test('success: getPotentialHistory', async () => {
      const count = 1000;
      const response = await api.getPotentialHistory(count);

      console.log(toString(response));
    });

    test('success: getPotentialHistory with date', async () => {
      const count = 1000;
      const response = await api.getPotentialHistory(count, {
        year: 2025,
        month: 7,
        day: 31,
      });
      console.log(toString(response));
    });

    test('success: getPotentialHistory with cursor', async () => {
      const count = 10;
      const response = await api.getPotentialHistory(count, {
        year: 2025,
        month: 7,
        day: 31,
      });
      const { nextCursor } = response; // 결과가 count 보다 많은 경우에만 nextCursor가 반환 됩니다.
      expect(nextCursor).toBeDefined();
      const nextResponse = await api.getPotentialHistory(count, nextCursor!);
      console.log(toString(nextResponse));
    });

    test('fail: getPotentialHistory with invalid date', async () => {
      try {
        await api.getPotentialHistory(1000, {
          year: 2024,
          month: 1,
          day: 24,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2024-01-25.');
        console.log(error.message);
      }
    });

    test('fail: getPotentialHistory with invalid cursor throw OPENAPI00003', async () => {
      const invalidCursor = 'invalid_cursor_123';
      try {
        await api.getPotentialHistory(1000, invalidCursor);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });

    test('fail: getPotentialHistory with excessive count throw OPENAPI00004', async () => {
      try {
        const count = 9999;
        await api.getPotentialHistory(count);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00004);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getCubeHistory', () => {
    test('success: getCubeHistory', async () => {
      const count = 1000;
      const response = await api.getCubeHistory(count);

      console.log(toString(response));
    });

    test('success: getCubeHistory with date', async () => {
      const count = 1000;
      const response = await api.getCubeHistory(count, {
        year: 2025,
        month: 10,
        day: 23,
      });
      console.log(toString(response));
    });

    test('success: getCubeHistory with cursor', async () => {
      const count = 10;
      const response = await api.getCubeHistory(count, {
        year: 2025,
        month: 10,
        day: 23,
      });
      const { nextCursor } = response; // 결과가 count 보다 많은 경우에만 nextCursor가 반환 됩니다.
      expect(nextCursor).toBeDefined();
      const nextResponse = await api.getCubeHistory(count, nextCursor!);
      console.log(toString(nextResponse));
    });

    test('fail: getCubeHistory with invalid cursor throw OPENAPI00003', async () => {
      const invalidCursor = 'invalid_cursor_123';
      try {
        await api.getCubeHistory(1000, invalidCursor);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });

    test('fail: getCubeHistory with excessive count throw OPENAPI00004', async () => {
      try {
        const count = 9999;
        await api.getPotentialHistory(count);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00004);
        console.log(error.errorCode, error.message);
      }
    });
  });
});

File: python/test/kms/history_api_test.py
import os
import sys
import unittest
import pytest

from datetime import datetime
from dotenv import load_dotenv

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../')))

from maplestory_openapi.api.common import MapleStoryApiException
from maplestory_openapi.api.kms import MapleStoryApi

load_dotenv()

api_key = os.environ['API_KEY_KMS'] # Your API Key
api = MapleStoryApi(api_key)

# History Information Retrieval

class TestGetStarforceHistory(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_starforce_history(self):
        count = 1000
        response = await api.get_starforce_history(count)
        print(response)

    async def test_success_get_starforce_history_with_date(self):
        count = 1000
        response = await api.get_starforce_history(count, date=datetime(2025, 9, 21))
        print(response)

    async def test_success_get_starforce_history_with_cursor(self):
        count = 10
        response = await api.get_starforce_history(count, date=datetime(2025, 9, 21))
        next_cursor = response.next_cursor
        assert next_cursor is not None
        next_response = await api.get_starforce_history(count, cursor=next_cursor)
        print(next_response)

    async def test_fail_get_starforce_history_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_starforce_history(1000, date=datetime(2023, 12, 26))
        assert 'You can only retrieve data after 2023-12-27' in str(e.value)
        print(e.value)

    async def test_fail_get_starforce_history_with_invalid_cursor_throw_OPENAPI00003(self):
        invalid_cursor = 'invalid_cursor_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_starforce_history(1000, cursor=invalid_cursor)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)

    async def test_fail_get_starforce_history_with_excessive_count_throw_OPENAPI00004(self):
        count = 9999
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_starforce_history(count)
        assert e.value.error_code == 'OPENAPI00004'
        print(e.value.error_code, e.value.message)


class TestGetPotentialHistory(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_potential_history(self):
        count = 1000
        response = await api.get_potential_history(count)
        print(response)

    async def test_success_get_potential_history_with_date(self):
        count = 1000
        response = await api.get_potential_history(count, date=datetime(2025, 7, 31))
        print(response)

    async def test_success_get_potential_history_with_cursor(self):
        count = 10
        response = await api.get_potential_history(count, date=datetime(2025, 7, 31))
        next_cursor = response.next_cursor
        assert next_cursor is not None
        next_response = await api.get_potential_history(count, cursor=next_cursor)
        print(next_response)

    async def test_fail_get_potential_history_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_potential_history(1000, date=datetime(2024, 1, 24))
        assert 'You can only retrieve data after 2024-01-25' in str(e.value)
        print(e.value)

    async def test_fail_get_potential_history_with_invalid_cursor_throw_OPENAPI00003(self):
        invalid_cursor = 'invalid_cursor_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_potential_history(1000, cursor=invalid_cursor)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)

    async def test_fail_get_potential_history_with_excessive_count_throw_OPENAPI00004(self):
        count = 9999
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_potential_history(count)
        assert e.value.error_code == 'OPENAPI00004'
        print(e.value.error_code, e.value.message)


class TestGetCubeHistory(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_cube_history(self):
        count = 1000
        response = await api.get_cube_history(count)
        print(response)

    async def test_success_get_cube_history_with_date(self):
        count = 1000
        response = await api.get_cube_history(count, date=datetime(2025, 10, 23))
        print(response)

    async def test_success_get_cube_history_with_cursor(self):
        count = 10
        response = await api.get_cube_history(count, date=datetime(2025, 10, 23))
        next_cursor = response.next_cursor
        assert next_cursor is not None
        next_response = await api.get_cube_history(count, cursor=next_cursor)
        print(next_response)

    async def test_fail_get_cube_history_with_invalid_cursor_throw_OPENAPI00003(self):
        invalid_cursor = 'invalid_cursor_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_cube_history(1000, cursor=invalid_cursor)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)

    async def test_fail_get_cube_history_with_excessive_count_throw_OPENAPI00004(self):
        count = 9999
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_potential_history(count)
        assert e.value.error_code == 'OPENAPI00004'
        print(e.value.error_code, e.value.message)


if __name__ == '__main__':
    unittest.main()
